### PR TITLE
fix: Return appropriate data type for time `mean` and `median`

### LINKED ISSUE #14470: `mean` and `median` for `pl.Time` should return a time
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

Part of #13599. `mean` and `median` either do not work, return null, or do not return the proper dtype, except in a groupby context.

```python
from datetime import time
import polars as pl

s = pl.Series([time(hour=1), time(hour=2), time(hour=15)])
df = pl.DataFrame({
    "key": [1, 1, 1],
    "s": s
})

### Series

print(s.mean())
print(s.median())
print(s.dt.mean())    # errors
print(s.dt.median())  # errors
```
```
21600000000000.0
21600000000000.0
AttributeError: type object 'Time' has no attribute 'time_unit'
```

### Frame

```python
print(
    df.select(
        col("s").mean().alias("mean"),
        # col("s").median().alias("median")  ## errors
    )
)
```
```
shape: (1, 1)
┌──────┐
│ mean │
│ ---  │
│ time │
╞══════╡
│ null │
└──────┘
```

### GroupBy

```python
print(
    df.group_by("key").agg(
        col("s").mean().alias("mean"),
        col("s").median().alias("median")
    )
)
```
```
shape: (1, 3)
┌─────┬──────────┬──────────┐
│ key ┆ mean     ┆ median   │
│ --- ┆ ---      ┆ ---      │
│ i64 ┆ time     ┆ time     │
╞═════╪══════════╪══════════╡
│ 1   ┆ 06:00:00 ┆ 02:00:00 │
└─────┴──────────┴──────────┘
```

### Log output

_No response_

### Issue description

a

### Expected behavior

a

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.7
Index type:           UInt32
Platform:             Linux-5.15.133.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python:               3.12.1 (main, Jan 31 2024, 09:51:46) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_manager:  0.9.0
cloudpickle:          3.0.0
connectorx:           <not installed>
deltalake:            0.15.2
fsspec:               2023.12.2
gevent:               23.9.1
hvplot:               0.9.2
matplotlib:           3.8.2
numpy:                1.26.4
openpyxl:             3.1.2
pandas:               2.2.0
pyarrow:              15.0.0
pydantic:             2.6.1
pyiceberg:            0.5.1
pyxlsb:               1.0.10
sqlalchemy:           2.0.25
xlsx2csv:             0.8.1
xlsxwriter:           3.1.9
```

</details>

### PR DESCRIPTION
**Pull request by mcrumiller**
Resolves #14470. Part of overall #13599.

This also moves one step closer to deprecating `Series.dt.mean` and `Series.dt.median`, as the base Series' mean and median now support **time** dtypes.

Using the example from the issue, all cases now properly output a duration:

```
06:00:00
02:00:00
06:00:00
02:00:00
shape: (1, 2)
┌──────────┬──────────┐
│ mean     ┆ median   │
│ ---      ┆ ---      │
│ time     ┆ time     │
╞══════════╪══════════╡
│ 06:00:00 ┆ 02:00:00 │
└──────────┴──────────┘
shape: (1, 2)
┌──────────┬──────────┐
│ mean     ┆ median   │
│ ---      ┆ ---      │
│ time     ┆ time     │
╞══════════╪══════════╡
│ 06:00:00 ┆ 02:00:00 │
└──────────┴──────────┘
```

### CODE REVIEW

File: py-polars/tests/unit/namespaces/test_datetime.py [hunk 1]
```diff
@@ -956,6 +959,9 @@ def test_weekday(time_unit: TimeUnit) -> None:
         "single_dur",
         "spread_even_dur",
         "spread_skewed_dur",
+        "single_time",
+        "spread_even_time",
+        "spread_skewed_time",
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/namespaces/test_datetime.py:964]: the
```
    if s.dtype == pl.Datetime:
        assert s.median() == expected_median
```
you have on line 973, and that now be extended to dtypes other than `Datetime`? Is it only `pl.Date` that's missing?

File: py-polars/tests/unit/namespaces/test_datetime.py [hunk 2]
```diff
@@ -1001,6 +1010,9 @@ def test_median(
         "single_duration",
         "spread_even_duration",
         "spread_skewed_duration",
+        "single_time",
+        "spread_even_time",
+        "spread_skewed_time",
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/namespaces/test_datetime.py:1015]: same

**Review by MarcoGorelli — COMMENTED** [review]
looks good, just some minor comments

File: py-polars/tests/unit/namespaces/test_datetime.py [hunk 3]
```diff
@@ -1032,6 +1044,27 @@ def test_datetime_mean_with_tu(values: list[datetime], expected_mean: datetime)
...
+)
+def test_datetime_median_with_tu(
+    values: list[datetime], expected_median: datetime
+) -> None:
+    assert pl.Series(values, dtype=pl.Duration("ms")).median() == expected_median
+    assert pl.Series(values, dtype=pl.Duration("ms")).dt.median() == expected_median
+    assert pl.Series(values, dtype=pl.Duration("us")).median() == expected_median
+    assert pl.Series(values, dtype=pl.Duration("us")).dt.median() == expected_median
+    assert pl.Series(values, dtype=pl.Duration("ns")).median() == expected_median
+    assert pl.Series(values, dtype=pl.Duration("ns")).dt.median() == expected_median
```
[the remark this answers is not in the document]
Comment (mcrumiller) [reply py-polars/tests/unit/namespaces/test_datetime.py:1065]: Yep. Thanks.

[the remark this answers is not in the document]
Comment (mcrumiller) [reply py-polars/tests/unit/namespaces/test_datetime.py:1065]: I've also cleaned up some of the IDs and added a `None` tests to the parametrization of a few.

**Comment by mcrumiller** [comment]
@MarcoGorelli I believe I've resolved all of your issues with minor test improvements:

* Parametrized time unit where applicable
* Added a few `None` cases to existing parametrizations
* Clarified the `id` of certain parametrizations
* Added an additional aggregate median test to match the existing aggregate mean test.

File: py-polars/tests/unit/namespaces/test_datetime.py [hunk 4]
```diff
@@ -1009,62 +1021,91 @@ def test_mean(
...
-    assert pl.Series(values, dtype=pl.Duration("us")).mean() == expected_mean
-    assert pl.Series(values, dtype=pl.Duration("us")).dt.mean() == expected_mean
-    assert pl.Series(values, dtype=pl.Duration("ns")).mean() == expected_mean
-    assert pl.Series(values, dtype=pl.Duration("ns")).dt.mean() == expected_mean
+@pytest.mark.parametrize("time_unit", ["ms", "us", "ns"])
+def test_datetime_mean_with_tu(
+    values: list[datetime], expected_mean: datetime, time_unit: str
+) -> None:
+    assert pl.Series(values, dtype=pl.Duration(time_unit)).mean() == expected_mean  # type:ignore[arg-type]
+    assert pl.Series(values, dtype=pl.Duration(time_unit)).dt.mean() == expected_mean  # type:ignore[arg-type]
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/namespaces/test_datetime.py:1044]: I think you can use `time:unit: TimeUnit`, and then you don't need the type: ignores

File: py-polars/tests/unit/namespaces/test_datetime.py [hunk 5]
```diff
@@ -1009,62 +1021,91 @@ def test_mean(
...
+        (
+            [datetime(2022, 1, 1), datetime(2022, 1, 2), datetime(2024, 5, 15)],
+            datetime(2022, 1, 2),
+        ),
+    ],
+    ids=["None_dt", "spread_skewed_dt"],
+)
+@pytest.mark.parametrize("time_unit", ["ms", "us", "ns"])
+def test_datetime_median_with_tu(
+    values: list[datetime], expected_median: datetime, time_unit: str
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/namespaces/test_datetime.py:1060]: same

File: py-polars/tests/unit/namespaces/test_datetime.py [hunk 6]
```diff
@@ -1009,62 +1021,91 @@ def test_mean(
...
+            [timedelta(days=1), timedelta(days=2), timedelta(days=15)],
+            timedelta(days=6),
+        ),
+    ],
+    ids=["None_dur", "spread_skewed_dur"],
 )
+@pytest.mark.parametrize("time_unit", ["ms", "us", "ns"])
 def test_duration_mean_with_tu(
-    values: list[timedelta], expected_mean: timedelta
+    values: list[timedelta], expected_mean: timedelta, time_unit: str
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/namespaces/test_datetime.py:1081]: same

File: py-polars/tests/unit/namespaces/test_datetime.py [hunk 7]
```diff
@@ -1009,62 +1021,91 @@ def test_mean(
...
+            [timedelta(days=1), timedelta(days=2), timedelta(days=15)],
+            timedelta(days=2),
+        ),
+    ],
+    ids=["None_dur", "spread_skewed_dur"],
 )
+@pytest.mark.parametrize("time_unit", ["ms", "us", "ns"])
 def test_duration_median_with_tu(
-    values: list[timedelta], expected_median: timedelta
+    values: list[timedelta], expected_median: timedelta, time_unit: str
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/namespaces/test_datetime.py:1100]: same

File: py-polars/tests/unit/namespaces/test_datetime.py (hunk 4 above)
Comment (mcrumiller) [reply py-polars/tests/unit/namespaces/test_datetime.py]: Much better, thanks!

**Review by MarcoGorelli — APPROVED** [review]
thanks @mcrumiller !

**Comment by mcrumiller** [comment]
Hi @stinodego, this one is relatively straightforward (virtually identical to #14376) and has been sitting in the queue for a while. Is there any chance you could take a look at this, so that I can get the `pl.Date` mean/median reworked with the latest main?

**Pull request merged**